src/packed_leaf.rs

```diff
@@ -1,132 +1,157 @@
-    #[arbitrary(with = arb_rwlock)]
-    pub hash: RwLock<Hash256>,
-    pub(crate) values: Vec<T>,
+pub struct PackedLeaf<T: Value> {
+    pub hash: Hash256,
+    pub length: u8,
+    _phantom: PhantomData<T>,
 }
 
 impl<T> Clone for PackedLeaf<T>
 where
-    T: TreeHash + Clone,
+    T: Value,
 {
     fn clone(&self) -> Self {
         Self {
-            hash: RwLock::new(*self.hash.read()),
-            values: self.values.clone(),
+            hash: self.hash,
+            length: self.length,
+            _phantom: PhantomData,
         }
     }
 }
 
-impl<T: TreeHash + Clone> PackedLeaf<T> {
-    pub fn tree_hash(&self) -> Hash256 {
-        let read_lock = self.hash.read();
-        let mut hash = *read_lock;
-        drop(read_lock);
-
-        if !hash.is_zero() {
-            return hash;
-        }
+impl<T: Value> PackedLeaf<T> {
+    pub fn length(&self) -> usize {
+        self.length as usize
+    }
 
-        let hash_bytes = hash.as_bytes_mut();
+    fn value_len() -> usize {
+        BYTES_PER_CHUNK / T::tree_hash_packing_factor()
+    }
 
-        let value_len = BYTES_PER_CHUNK / T::tree_hash_packing_factor();
-        for (i, value) in self.values.iter().enumerate() {
-            hash_bytes[i * value_len..(i + 1) * value_len]
-                .copy_from_slice(&value.tree_hash_packed_encoding());
+    pub fn get(&self, index: usize) -> Option<&T> {
+        if index >= self.length() {
+            return None;
         }
+        let hash_base_ptr: *const Hash256 = &self.hash;
+        let base_ptr: *const T = hash_base_ptr as *const T;
+        let elem_ptr: *const T = unsafe { base_ptr.add(index) };
+        Some(unsafe { &*elem_ptr })
+    }
 
-        *self.hash.write() = hash;
-        hash
+    pub fn tree_hash(&self) -> Hash256 {
+        self.hash
     }
 
     pub fn empty() -> Self {
         PackedLeaf {
-            hash: RwLock::new(Hash256::zero()),
-            values: Vec::with_capacity(T::tree_hash_packing_factor()),
+            hash: Hash256::zero(),
+            length: 0,
+            _phantom: PhantomData,
         }
     }
 
     pub fn single(value: T) -> Self {
-        let mut values = Vec::with_capacity(T::tree_hash_packing_factor());
-        values.push(value);
+        let mut hash = Hash256::zero();
+        let hash_bytes = hash.as_bytes_mut();
+
+        let value_len = Self::value_len();
+        hash_bytes[0..value_len].copy_from_slice(&value.as_ssz_bytes());
 
         PackedLeaf {
-            hash: RwLock::new(Hash256::zero()),
-            values,
+            hash,
+            length: 1,
+            _phantom: PhantomData,
         }
     }
 
     pub fn repeat(value: T, n: usize) -> Self {
         assert!(n <= T::tree_hash_packing_factor());
+
+        let mut hash = Hash256::zero();
+        let hash_bytes = hash.as_bytes_mut();
+
+        let value_len = Self::value_len();
+
+        for (i, value) in vec![value; n].iter().enumerate() {
```

Comment:
Ooh, meant to flag this too. We can avoid allocating a vec here by using `std::iter::repeat(value).take(n)`.